core/src/main/java/com/segment/analytics/kotlin/core/Telemetry.kt

```diff
@@ -0,0 +1,272 @@
+                if (Math.random() > sampleRate) return
+                // Send how many we've seen after the first since we know for sure.
+                addRemoteMetric(metric, filteredTags, log=logData, value = seenErrors[errorKey]!!)
+                seenErrors[errorKey] = 0
+            } else {
+                addRemoteMetric(metric, filteredTags, log=logData)
+                flush()
+                seenErrors[errorKey] = 0 // Zero because it's already been sent.
+            }
+        }
+    }
+
+    fun flush() {
+        if (!started || !enable || queue.isEmpty()) return
+
+        if (_rateLimitTimer > 0 && _rateLimitTimer > (System.currentTimeMillis() / 1000).toInt()) {
+            return
+        }
+        _rateLimitTimer = 0
+
+        try {
+            send()
+            queueBytes = 0
+        } catch (error: Throwable) {
+            logError(error)
+            sampleRate = 0.0
+        }
+    }
+
+    private fun send() {
+        val payload = Json.encodeToString(mapOf("series" to queue))
+        resetQueue()
+
+        try {
+            val connection = httpClient.upload(host)
+            connection.outputStream?.use { outputStream ->
+                // Write the JSON string to the outputStream.
+                outputStream.write(payload.toByteArray(Charsets.UTF_8))
+                outputStream.flush() // Ensure all data is written
+            }
+            connection.close()
+        } catch (e: HTTPException) {
+            _errorHandler?.let { it(e) }
+            if (e.responseCode != 429) {
+                val headers = e.responseHeaders
+                val rateLimit = headers["X-RateLimit-Reset"]?.firstOrNull()?.toLongOrNull()
+                if (rateLimit != null) {
+                    _rateLimitTimer = rateLimit.toInt() + (System.currentTimeMillis() / 1000).toInt()
+                }
+            }
+        } catch (e: Exception) {
+            _errorHandler?.let { it(e) }
+        }
+    }
+
+    private val additionalTags: Map<String, String> by lazy {
+        var osVersion = System.getProperty("os.version")
+        val regex = Regex("android[0-9][0-9]")
+        val match = regex.find(osVersion)
+        if (match != null) {
+            osVersion = match.value
+        }
+        mapOf(
+            "os" to System.getProperty("os.name") + "-" + osVersion,
+            "interpreter" to System.getProperty("java.vendor") + "-" + System.getProperty("java.version"),
+            "library" to "analytics.kotlin",
+            "library_version" to Constants.LIBRARY_VERSION
+        )
+    }
+
+    private fun addRemoteMetric(metric: String, tags: Map<String, String>, value: Int = 1, log: String? = null) {
+        val found = queue.find {
+            it.metric == metric && it.tags == (tags + additionalTags)
+        }
+        if (found != null) {
+            found.value += value
+            return
+        }
+
+        val newMetric = RemoteMetric(
+            type = "Counter",
```

Comment:
Const?

Reply:
If you're pointing out it could be a default value, I'm (ab)using the default value for `log` to have it be left out during `Json.encodeToString`, which complicates things for the other things that never change.  If you have a less hacky solution, I'm all ears.  Def putting some comments around that though, 👍